lib/ecs_deploy/auto_scaler/instance_fluctuation_manager.rb

```diff
@@ -0,0 +1,165 @@
+require "aws-sdk-autoscaling"
+require "aws-sdk-ec2"
+require "aws-sdk-ecs"
+
+module EcsDeploy
+  module AutoScaler
+    class InstanceFluctuationManager
+      attr_reader :logger
+
+      def initialize(region:, cluster:, cluster_to_asg:, desired_capacity:, logger:)
+        @region = region
+        @cluster = cluster
+        @cluster_to_asg = cluster_to_asg
+        @desired_capacity = desired_capacity
+        @logger = logger
+
+        unless cluster_to_asg.key?(cluster)
+          raise ArgumentError, "Unknown cluster: #{cluster}"
+        end
+
+        @asg_name = cluster_to_asg[cluster]
+      end
+
+      def increase
+        asg = as_client.describe_auto_scaling_groups(auto_scaling_group_names: [@asg_name]).auto_scaling_groups.first
+
+        @logger.info("Increase desired capacity of #{@asg_name}: #{asg.desired_capacity} => #{asg.max_size}")
+        as_client.update_auto_scaling_group(auto_scaling_group_name: @asg_name, desired_capacity: asg.max_size)
+
+        # Run in background because increasing instances may take time
+        Thread.new do
+          loop do
+            cluster = ecs_client.describe_clusters(clusters: [@cluster]).clusters.first
+            instance_count = cluster.registered_container_instances_count
+            if instance_count == asg.max_size
+              @logger.info("Succeeded to increase instance count!")
+              break
+            end
+            @logger.info("Current registered instance count: #{instance_count}")
+            sleep 5
+          end
+        end
+      end
+
+      def decrease
+        asg = as_client.describe_auto_scaling_groups(auto_scaling_group_names: [@asg_name]).auto_scaling_groups.first
+
+        decrease_count = asg.desired_capacity - @desired_capacity
+        if decrease_count <= 0
+          @logger.info("The capacity is already #{asg.desired_capacity}")
+          return
+        end
+        @logger.info("Decrease desired capacity of #{@asg_name}: #{asg.desired_capacity} => #{@desired_capacity}")
+
+        container_instance_arns = ecs_client.list_container_instances(
+          cluster: @cluster
+        ).container_instance_arns
+        container_instances = ecs_client.describe_container_instances(
+          cluster: @cluster,
+          container_instances: container_instance_arns
+        ).container_instances
+
+        region_to_container_instances = container_instances.sort_by {|ci| - ci.running_tasks_count }.group_by do |ci|
+          ci.attributes.find {|attribute| attribute.name == "ecs.availability-zone" }.value
+        end
+        if region_to_container_instances.empty?
+          @logger.info("There are no instances to terminate.")
+          return
+        end
+        target_container_instances = decrease_count.times.map do |i|
```

Comment:
[must] This logic can terminate instances unexpectedly if AZRebalance is enabled.
For example, there are two instances in ap-northeast-1c while there are no instances in ap-northeast-1a after instances are detached in the following case:

Instance | AZ  | Task Count
-------|----------------|---------
i-0001 | ap-northeast-1c | 1
i-0002 | ap-northeast-1c | 1
i-0003 | ap-northeast-1a | 0

It results in launching a new instance in ap-northeast-1a and terminating an instance in ap-northeast-1c.

Reply:
Do you mean that we should call Aws::AutoScaling::Client#suspend_processes and Aws::AutoScaling::Client#resume_processes around terminating tasks and detaching instances?

Pseudocode:
```ruby
def decrease
  target_container_instances = detect_target_container_instances
  begin
    as_client.suspend_processes(auto_scaling_group_name: "xxx", scaling_processes: ["AZRebalance"])
    target_container_instances.each do |arn|
      stop_tasks(arn)
    end
    instance_ids = target_container_instances.map(&:ec2_instance_id)
    terminate_instances(instance_ids)
  ensure
    as_client.resume_processes(auto_scaling_group_name: "xxx", scaling_processes: ["AZRebalance"])
  end
end
```

See also: https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/AutoScaling/Client.html#resume_processes-instance_method

Reply:
No, I mean we should ensure detaching instances in an AZ which has the most number of instances.
For examples, we should detach an instance in ap-northeast-1c in the following case:

Instance | AZ  | Task Count
-------|----------------|---------
i-0001 | ap-northeast-1c | 1
i-0002 | ap-northeast-1c | 1
i-0003 | ap-northeast-1a | 0

Reply:
https://github.com/reproio/ecs_deploy/pull/52 might be helpful.
What this PR tries to realize is more complicated because we have to drain an instance in ap-northeast-1c in advance in the example I mentioned before.

Reply:
Fixed in cc46897